### PR TITLE
[LLD] dtNeeded name should consistent with soNames

### PR DESCRIPTION
**Pull request by CoTinker**
If dtNeeded name is inconsistent with soNames, lld can not report error right.
### demo
```
#--- def.s

#--- ref.s
.globl fun
fun:
  bl foo@PLT

#--- main.s
.globl _start
_start:
  bl fun@PLT


$ llvm-mc -filetype=obj -triple=aarch64 main.s -o main.o
$ llvm-mc -filetype=obj -triple=aarch64 def.s -o def.o && ld.lld -shared def.o -o def.so
$ llvm-mc -filetype=obj -triple=aarch64 ref.s -o ref.o && ld.lld -shared ref.o -o ref.so ./def.so
```

### can report error
$ ld.lld main.o ref.so ./def.so
```
ld.lld: error: undefined reference due to --no-allow-shlib-undefined: foo
>>> referenced by ref.so
```
dtNeeded name = `./def.so`, soName = `./def.so`, consistent.
### can not report error
$ ld.lld main.o ref.so def.so
dtNeeded name = `./def.so`, soName = `def.so`, inconsistent.

We can see that if dtNeeded name is inconsistent with soNames, lld can not report error right.
https://github.com/llvm/llvm-project/blob/4594d5bb3ac6772bb20e429bbb04842ef6eaea35/lld/ELF/Writer.cpp#L2019-L2036

### CODE REVIEW

**Comment by CoTinker** [comment]
@MaskRay

**Comment by smithp35** [comment]
I'm finding your examples a bit hard to follow. Can you clarify a few things:
* Did you mean have.c to just be a declaration? There is no foo in the dynamic symbol.
* What is ususe.c? Is it just use.c renamed?
* Is -o exe/libuseo a mistake -luse won't pick that up as it looks for libuse.a or libuse.so
* Please can you write a test for your patch? Look for examples in lld/ELF/tests

**Comment by CoTinker** [comment]
> I'm finding your examples a bit hard to follow. Can you clarify a few things:
> 
> * Did you mean have.c to just be a declaration? There is no foo in the dynamic symbol.
> * What is ususe.c? Is it just use.c renamed?
> * Is -o exe/libuseo a mistake -luse won't pick that up as it looks for libuse.a or libuse.so
> * Please can you write a test for your patch? Look for examples in lld/ELF/tests

1. YES，the foo function is declared but not implemented.
2. That's my fault, I wrote it wrong.
3. No, actually in case 3, dtneeded name is exe/libhave.so, but soName is libhave.so in -lhave, so there are inconsistent. In case 4, dtneeded name is libhave.so, but soName is exe/libhave.so, alse inconsistent.
4. I'll try it.

**Comment by MaskRay** [comment]
I suspect that you have an incorrect use case, but the case is not clearly described and I cannot verify.
This inconcistency what we will get when a shared object does not have `DT_SONAME` (https://maskray.me/blog/2020-11-15-explain-gnu-linker-options#sonamename)

FYI I place all needed files in a directory and run `cd dir; mkreproduce . > /tmp/rep.sh` to create `cat <<eof` and `echo` instructions that other people can easily reproduce. (e.g. https://maskray.me/blog/2020-11-26-all-about-symbol-versioning#version-script)
```
#!/bin/zsh
dry_run=
if [[ $1 == --dry-run ]]; then dry_run=1; shift; fi

f() {
  for f in $1/*(ND.^x); do
    [[ ${f##*/} =~ '^\.' ]] && continue
    if [[ -n $dry_run ]]; then
      echo $f
    elif [[ $(wc -l < $f) == 1 ]]; then
      echo "echo '$(<$f)' > $f"
    else
      if [[ ${f##*/} == Makefile ]]; then
        printf '%s\n' "sed 's/^        /\\t/' > $f <<'eof'"
      else
        echo "cat > $f <<'eof'"
      fi
      cat $f
      echo eof
    fi
  done
  for d in $1/*(ND/); do
    [[ $d =~ '^\.' ]] && continue
    echo mkdir -p $d
    f $d
  done
}

f ${1:-.}
```

```
clang case/use.c -o exe/libuse.so -lhave -Lexe --shared -fPIC
```

The `-o libuse.so` instruction does not mention how you switched the directory, so the example is difficult to follow.

**Comment by CoTinker** [comment]
> I suspect that you have an incorrect use case, but the case is not clearly described and I cannot verify. This inconsistency fault is likely at the user side when a shared object does not have `DT_SONAME` (https://maskray.me/blog/2020-11-15-explain-gnu-linker-options#sonamename)
> 
> FYI I place all needed files in a directory and run `cd dir; mkreproduce . > /tmp/rep.sh` to create `cat <<eof` and `echo` instructions that other people can easily reproduce. (e.g. https://maskray.me/blog/2020-11-26-all-about-symbol-versioning#version-script is an example using the script to prepare for input files)
> 
> ```
> #!/bin/zsh
> dry_run=
> if [[ $1 == --dry-run ]]; then dry_run=1; shift; fi
> 
> f() {
>   for f in $1/*(ND.^x); do
>     [[ ${f##*/} =~ '^\.' ]] && continue
>     if [[ -n $dry_run ]]; then
>       echo $f
>     elif [[ $(wc -l < $f) == 1 ]]; then
>       echo "echo '$(<$f)' > $f"
>     else
>       if [[ ${f##*/} == Makefile ]]; then
>         printf '%s\n' "sed 's/^        /\\t/' > $f <<'eof'"
>       else
>         echo "cat > $f <<'eof'"
>       fi
>       cat $f
>       echo eof
>     fi
>   done
>   for d in $1/*(ND/); do
>     [[ $d =~ '^\.' ]] && continue
>     echo mkdir -p $d
>     f $d
>   done
> }
> 
> f ${1:-.}
> ```
> 
> ```
> clang case/use.c -o exe/libuse.so -lhave -Lexe --shared -fPIC
> ```
> 
> The `-o libuse.so` instruction does not mention how you switched the directory, so the example is difficult to follow.

Thank you, the core problem is probably shared object does not have DT_SONAME.
